src/libraries/System.Runtime.Numerics/src/System/Numerics/BigNumber.cs

```diff
@@ -494,35 +494,243 @@ private static bool HexNumberToBigInteger(ref BigNumberBuffer number, out BigInt
-                }
+                    totalDigitCount = Math.Min(number.digits.Length - 1, numberScale);
+                    int bufferSize = (totalDigitCount + MaxPartialDigits - 1) / MaxPartialDigits;
 
-                if (partialDigitCount > 0)
-                {
-                    MultiplyAdd(ref currentBuffer, s_uint32PowersOfTen[partialDigitCount], partialValue);
+                    Span<uint> buffer = new uint[bufferSize];
+
+                    int bufferInd = buffer.Length - 1;
+                    uint currentBlock = 0;
+                    int shiftUntil = (totalDigitCount - 1) % MaxPartialDigits;
+                    int remainingIntDigitCount = totalDigitCount;
+                    foreach (ReadOnlyMemory<char> digitsChunk in number.digits.GetChunks())
+                    {
+                        var digitsChunkSpan = digitsChunk.Span;
+                        ReadOnlySpan<char> intDigitsSpan = digitsChunkSpan.Slice(0, Math.Min(remainingIntDigitCount, digitsChunkSpan.Length));
+
+                        for (int i = 0; i < intDigitsSpan.Length; i++)
+                        {
+                            char digitChar = intDigitsSpan[i];
+                            Debug.Assert(char.IsDigit(digitChar));
+                            currentBlock = currentBlock * 10 + digitChar - '0';
+                            if (shiftUntil == 0)
+                            {
+                                buffer[bufferInd] = currentBlock;
+                                currentBlock = 0;
+                                bufferInd--;
+                                shiftUntil = MaxPartialDigits;
+                            }
+                            shiftUntil--;
+                        }
+                        remainingIntDigitCount -= intDigitsSpan.Length;
+
+                        ReadOnlySpan<char> fracDigitsSpan = digitsChunkSpan.Slice(intDigitsSpan.Length);
+                        for (int i = 0; i < fracDigitsSpan.Length; i++)
+                        {
+                            char digitChar = fracDigitsSpan[i];
+                            if (digitChar == '\0')
+                            {
+                                break;
+                            }
+                            if (digitChar != '0')
+                            {
+                                result = default;
+                                return false;
+                            }
+                        }
+                    }
+                    Debug.Assert(bufferInd == -1);
+
+                    unsafe
+                    {
+                        Span<uint> newBuffer = new uint[bufferSize];
+
+                        arrayFromPool = ArrayPool<int>.Shared.Rent(1);
+                        Span<uint> multiplier = MemoryMarshal.Cast<int, uint>(arrayFromPool);
+                        multiplier[0] = TenPowMaxPartial;
+
+                        int blockSize = 1;
+                        while (true)
```

Comment:
nit: I'm not a huge fan of `while (true)` loops as it makes the exit condition "non obvious"

In particular, I think this could probably be something like:
```csharp

do
{
    newBuffer.Clear();
    int[]? arrayToReturn = arrayFromPool;

    arrayFromPool = ArrayPool<int>.Shared.Rent(blockSize);
    Span<uint> newMultiplier = MemoryMarshal.Cast<int, uint>(arrayFromPool);
    newMultiplier.Clear();

    fixed (uint* mulPtr = multiplier, newMulPtr = newMultiplier)
    {
        BigIntegerCalculator.Square(mulPtr, blockSize / 2, newMulPtr, blockSize);
    }

    multiplier = newMultiplier;

    if (arrayToReturn is not null)
    {
        ArrayPool<int>.Shared.Return(arrayToReturn);
    }


    // L658...L702: https://github.com/dotnet/runtime/pull/55121/files#diff-83081f6bd1123c4813c96a452ead0a065d670bbf44ed85023f772253eaf7d2c5R658-R702
}
while (bufferSize > blockSize);
```

It adds some minimal additional up-front cost, but makes it easier too follow the general flow of logic. The up front cost (what is currently the trailing/would become the leading logic) could likely be reduced as well with some minimal changes.

Reply:
I also didn't want to use the `while (true)` statement. However, I decided it was best to do so in this case. I came up with two different solutions.
## first solution
```cs
for (blockSize = 1; blockSize < bufferSize; blockSize *= 2, UpdateMultiplier(ref multiplier))
{
    // https://github.com/dotnet/runtime/pull/55121/files#diff-83081f6bd1123c4813c96a452ead0a065d670bbf44ed85023f772253eaf7d2c5R658-R702
    newBuffer.Clear();
}
void UpdateMultiplier(ref Span<uint> multiplier)
{
    // https://github.com/dotnet/runtime/pull/55121/files#diff-83081f6bd1123c4813c96a452ead0a065d670bbf44ed85023f772253eaf7d2c5R709-R722
}
```
This is the most intuitive one, but it adds additional `UpdateMultiplier` calls compare to the original one. This costs a not negligible time since the last `UpdateMultiplier` generates a large `multiplier`. 

## second solution
```cs
bool isFirstIteration = true;
do
{
    newBuffer.Clear();
    int[]? arrayToReturn = arrayFromPool;

    arrayFromPool = ArrayPool<int>.Shared.Rent(blockSize);
    Span<uint> newMultiplier = MemoryMarshal.Cast<int, uint>(arrayFromPool);
    newMultiplier.Clear();

    if (isFirstIteration)
    {
         newMultiplier[0] = TenPowMaxPartial;
         isFirstIteration = false;
    }
    else
    {
        fixed (uint* mulPtr = multiplier, newMulPtr = newMultiplier)
        {
            BigIntegerCalculator.Square(mulPtr, blockSize / 2, newMulPtr, blockSize);
        }
    }

    multiplier = newMultiplier;

    if (arrayToReturn is not null)
    {
        ArrayPool<int>.Shared.Return(arrayToReturn);
    }

    // L658...L702: https://github.com/dotnet/runtime/pull/55121/files#diff-83081f6bd1123c4813c96a452ead0a065d670bbf44ed85023f772253eaf7d2c5R658-R702
}
while (bufferSize > blockSize);
```
this solves the above problem, but for me, it is harder to understand what is going on than the original code.

Reply:
I was thinking it could be done more like the bottom case but without needing to differentiate on `isFirstIteration`, it would increase the cost slightly of the first loop but simplify the logic and number of inner-loop conditions that need to otherwise be checked

Reply:
I'm fine with leaving it as is given the complexity around changing it and no clear solution.